### PR TITLE
Update Registry to respect provider aliases

### PR DESCRIPTION
**Pull request by EvanBoyle**
# Description

This change makes the provider registry aware of aliases. This allows provider aliases to work properly, ensuring that resources that use aliased providers don't get replaced unnecessarily.

Fixes https://github.com/pulumi/pulumi/issues/3979

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

### CODE REVIEW

**Comment by pgavlin** [comment]
It feels like this will behave subtly differently than other aliases, and I'm not sure that's what we want.

For other resource types, an alias just means "consider this resource to have the same identity as the resource in the old state". For providers, that identity influences a couple things:

1. Which provider instance is used for a particular resource, which is managed by the provider map
2. Whether or not resources that refer to the provider will be considered as requiring replacement due to a change in their provider

I think that your changes address (1)--updating the provider map ensures that resources in the old state will treat the new provider as having the same identity as their provider if it has been aliased--but I'm not sure that it addresses (2). I think that the referenced issue requires that (2) is fixed. Can you add a test for that scenario?

**Comment by EvanBoyle** [comment]
Just tested (2) out and you are correct, altering the provider results in it getting replaced but not the resources that depend on it. 

Do have any suggestions on what layer to look at for addressing (2)?

**Comment by joeduffy** [comment]
Hey @EvanBoyle I was super excited to see you working on a fix for this. Any chance it will be landing soonish?

**Comment by EvanBoyle** [comment]
Update, it looks like the incorrect behavior for (2) is due to https://github.com/pulumi/pulumi-terraform-bridge/issues/244

I'll write some lifecycle tests for (2) to verify.

**Comment by EvanBoyle** [comment]
@pgavlin I've added a passing test that covers those two scenarios, would you mind taking another look?

File: pkg/resource/deploy/providers/registry.go [hunk 1]
```diff
@@ -254,6 +254,18 @@ func (r *Registry) Check(urn resource.URN, olds, news resource.PropertyMap,
 	return inputs, nil, nil
 }
 
+// RegisterAliases takes the computed alias map, and updates the registry such that the new URNSs
+// point to the alised providers
```
Comment (pgavlin) [inline pkg/resource/deploy/providers/registry.go:258]:
```suggestion
// RegisterAliases takes the computed alias map and updates the registry such that the new URNs
// point to the aliased providers
```

File: pkg/resource/deploy/step_generator.go [hunk 2]
```diff
@@ -233,6 +233,9 @@ func (sg *stepGenerator) generateSteps(event RegisterResourceEvent) ([]Step, res
 		}
 	}
 
+	// register aliases with the provider registry
+	sg.deployment.providers.RegisterAliases(sg.aliased)
```
Comment (pgavlin) [inline pkg/resource/deploy/step_generator.go:237]: This is overkill: it's re-registering every alias after each resource registration. I think that we can move this into the code above that does the usual alias processing and change it s.t. it only handles a single alias at a time.

**Review by pgavlin — CHANGES_REQUESTED** [review]

File: pkg/resource/deploy/providers/registry.go [hunk 3]
```diff
@@ -254,6 +254,15 @@ func (r *Registry) Check(urn resource.URN, olds, news resource.PropertyMap,
 	return inputs, nil, nil
 }
 
+// RegisterAlias updates the registry such that the new URN points to the aliased provider
+func (r *Registry) RegisterAlias(alias, urn resource.URN) {
+	for ref, prov := range r.providers {
+		if ref.urn == alias {
+			r.setProvider(mustNewReference(urn, ref.id), prov)
+		}
+	}
```
Comment (pgavlin) [inline pkg/resource/deploy/providers/registry.go:263]: In principle we shouldn't need to do this. The caller can know whether or not the aliased resource is a provider by checking the type portion of the URN. I'd prefer that we be more precise here and do those checks in the step generator s.t. the contract for this method can be "alias this provider to this URN, and a provider with the alias target must exist". I think that will make the behavior clearer in both cases.

**Review by pgavlin — APPROVED** [review]

**Pull request merged**